Meme/Meme/NetworkClient.swift

```diff
@@ -9,47 +9,35 @@
 import Foundation
 
 protocol MemeClient {
-    func getMeme(from template: Template) -> Meme?
+    func getTemplates(_ data: Data?) -> [Template]?
 }
 
 class NetworkClient: MemeClient {
 
-    fileprivate func getTemplates(_ data: Data?) -> [Template]? {
-        guard let data = data,
-            let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String:String] else {
-                print("Error deserializing template data");
-                return nil
-        }
-        return jsonResponse.compactMap({ return Template.init(name: $0.0, templateURLString: $0.1)})
-    }
-
-    func requestTemplates(completion: @escaping ([Template]) -> Void)  {
+    func requestTemplates(completion: @escaping (Result<[Template], Error>) -> Void)  {
         let session = URLSession.shared
         let urlBuilder = MemeEndpointBuilder()
         let url = urlBuilder.endpointURL(.template)
         let request = URLRequest(url:url)
         let task: URLSessionDataTask = session.dataTask(with: request) { (data, response, error) -> Void in
-            guard let templates = self.getTemplates(data) else { return }
-            completion(templates)
+            guard let templates = self.getTemplates(data) else {
+                if let error = error {
+                    completion(.failure(error))
+                }
+                return
+            }
+            completion(.success(templates))
         }
         task.resume()
     }
 
-    func requestMemes(completion: @escaping ([Meme]?) -> Void) {
-        requestTemplates { [weak self] templates in
-            completion(templates.compactMap({ return self?.getMeme(from: $0)}))
-        }
-    }
-
-    func getMeme(from template: Template) -> Meme? {
-        let urlBuilder = MemeEndpointBuilder()
-        guard let imageAlias = template.imageAlias,
-            let  imageURL = urlBuilder.imageURL(with: imageAlias) else {
-                print("Error constructing Meme model")
+    internal func getTemplates(_ data: Data?) -> [Template]? {
```

Comment:
no need to add `internal`. `Data` should be qualified by this point. Also because this function can fail at the serialization step we should return `Result<[Template], RetrievalError>` that way we don't return nil.

Reply:
Fair point, about the `Data`. The `internal` qualifier was something that Xcode suggested since I moved the `getTemplates` function to the `MemeClient` protocol. Prior to that it was defined as a `fileprivate func`. What is the advantage/disadvantage of `internal`

Reply:
So `internal` is the default access level. The five in swift are
- open 
- public 
- internal 
- fileprivate 
- private

Internal gives the module access to the func/obj.